src/docs/binding-generics.md

```diff
@@ -437,3 +437,191 @@ We could also consider solving this on the runtime level.
 ### Generic classes
 
 TBD
+
+## Protocol constraints
+
+This section describes how to project protocol constraints on generic type parameters from Swift to C\#. The goal here is not to describe full projections of protocols; for that, please refer to the [binding protocols doc](binding-protcols.md).
+
+A Swift protocol can be mapped to C\# as an interface. Then Swift generic parameters with protocol constraint can be mapped into C\# as generic parameters with interface constraint. This should work for both functions and types.
+
+### Protocol constraints on functions
+
+Consider the following simple example in Swift:
+
+```swift
+public protocol Printable {
+ func printMe()
+}
+
+public func printPrintable<T: Printable>(data: T) {
+ data.printMe()
+}
+
+// Equivalent to: public func printPrintable<T>(data: T) where T : Printable
+```
+
+When a function is declared with a generic argument, an extra implicit argument is added to the end representing the type metadata of the generic argument (see above). If that generic argument is constrained by one or more protocols, the type metadata is followed by a pointer to the protocol witness table for each protocol. In the case of multiple protocols, e.g. T: P3 & P1 & P2, the protocol witness tables are ordered lexically by fully qualified name of the protocol. For detailed description of Metadata accessor refer to [runtime metadata doc](runtime-metadata.md).
+
+This can be projected into C\# as follows:
+
+```csharp
+public interface IPrintableProtocol
+{
+    public void PrintMe();
+    public abstract static NativeHandle GetIPrintableProtocolWitnessTable();
+
+}
+
+public static void PrintPrintable<T>(T data) where T : ISwiftObject, IPrintableProtocol
+{
+    var metadata = TypeMetadata.GetTypeMetadataOrThrow<T>();
+    var protocolWitnessTable = T.GetIPrintableProtocolWitnessTable();
+    PInvokePrintPrintable(data.GetPayload(), metadata, protocolWitnessTable);
+}
```

Comment:
Sort of related question - what type are we going to generate this method into? In theory it could go onto the interface itself, but then it's polluting the C# consumer view. If it goes to a different type then we need to be able to deduce that type from the interface. I know it's a sort of implementation detail, but I'm curious if we're thinking about sharing that type across multiple bindings as well.
For example let's say `IPrintableProtocol` is defined in a PrintBase framework. We project that to PrintBase.dll (and the nuget around it). Then when I'm in my app and I'm going to define a `PrintableDocument` struct/class and it implemented `IPrintableProtocol` in its implementation it will need access to this helper - would we call that helper from the `PrintBase.dll` or would we generate it again into the app? (And sorry, I'm stealing your design discussion again - for feel free to kick me out into a separate issue ;-))

Reply:
> what type are we going to generate this method into? In theory it could go onto the interface itself, but then it's polluting the C# consumer view. If it goes to a different type then we need to be able to deduce that type from the interface. 

In this design I put the method on the interface itself which is the easiest option but I agree that it will pollute the interface. I am not sure how big of an issue is this pollution (any swift type we are going to project will have a public `Metadata` method coming from `ISwiftObject` anyway.

If we wanted to avoid pollution though, we could revert to stripping this information from the type itself and reverting to some sort of mapping.

>For example let's say IPrintableProtocol is defined in a PrintBase framework. We project that to PrintBase.dll (and the nuget around it). Then when I'm in my app and I'm going to define a PrintableDocument struct/class and it implemented IPrintableProtocol in its implementation it will need access to this helper - would we call that helper from the PrintBase.dll or would we generate it again into the app?

First, protocol witness table (PWT) is generated for each pair (type, protocol), so if you defined a new type which extends this protocol on the csharp side only, we would need to go and generate the appropriate `PrintableDocument` on the Swift side so that a new PWT is correctly generated. And then, as far as I undersatand, we will be forced to either add a new `Get<ProtocolName>ProtocolWitnessTable` on the user defined type or add a new entry (TypeName, ProtocolName) -> PWT to some map.

Reply:
I gave it more thought. Just to correct myself - we can allow user to implement the interface on the c# side only without a need to to generate the swift side, I think Steve has a proposal in `binding-protocols` doc for that. So when dealing with this scenario we wont need / have a PWT for such combination (there will be no Swift type). We will have some proxy which will decided whether a type is a Swift type and has a PWT (and has to call swift) or is a C# side only type and needs to be handled on C# side exclusively. 

With that in mind I think we should should **not** put `GetProtocolWitnessTable` on the interface itself as some of types implementing it will not have anything meaningful to return there. We could have a static function which does map lookup.

Reply:
Not sure I fully understand - but I would prefer we avoid lookups unless necessary. If we know what needs to happen at build time (projection generation time), we should resolve it there.

As an aside, I'm curious what is the scenario where I declare an interface in C#, want it projected to Swift, but there's no such protocol in Swift. It doesn't feel like something we need right now, so maybe we can live without it...